streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskMovement.java

```diff
@@ -16,128 +16,94 @@
-     * {@code max.warmup.replicas}. Instead they will be counted against that task's total {@code num.standby.replicas}.
-     *
-     * @param statefulActiveTaskAssignment the initial, state constrained assignment, with the source clients
-     * @param balancedStatefulActiveTaskAssignment the final, balanced assignment, with the destination clients
-     * @return list of the task movements from statefulActiveTaskAssignment to balancedStatefulActiveTaskAssignment
+     * @return whether any warmup replicas were assigned
      */
-    static List<TaskMovement> getMovements(final Map<UUID, List<TaskId>> statefulActiveTaskAssignment,
-                                           final Map<UUID, List<TaskId>> balancedStatefulActiveTaskAssignment,
-                                           final Map<TaskId, SortedSet<UUID>> tasksToCaughtUpClients,
-                                           final Map<UUID, ClientState> clientStates,
-                                           final Map<TaskId, Integer> tasksToRemainingStandbys,
-                                           final int maxWarmupReplicas) {
-        if (statefulActiveTaskAssignment.size() != balancedStatefulActiveTaskAssignment.size()) {
-            throw new IllegalStateException("Tried to compute movements but assignments differ in size.");
-        }
+    static boolean assignTaskMovements(final Map<UUID, List<TaskId>> statefulActiveTaskAssignment,
+                                       final Map<TaskId, SortedSet<UUID>> tasksToCaughtUpClients,
+                                       final Map<UUID, ClientState> clientStates,
+                                       final Map<TaskId, Integer> tasksToRemainingStandbys,
+                                       final int maxWarmupReplicas) {
+        boolean warmupReplicasAssigned = false;
+
+        final ValidClientsByTaskLoadQueue clientsByTaskLoad =
+            new ValidClientsByTaskLoadQueue(
+                clientStates,
+                (client, task) -> taskIsCaughtUpOnClient(task, client, tasksToCaughtUpClients)
+            );
 
-        final Map<TaskId, UUID> taskToDestinationClient = new HashMap<>();
-        for (final Map.Entry<UUID, List<TaskId>> clientEntry : balancedStatefulActiveTaskAssignment.entrySet()) {
-            final UUID destination = clientEntry.getKey();
-            for (final TaskId task : clientEntry.getValue()) {
-                taskToDestinationClient.put(task, destination);
+        final SortedSet<TaskMovement> taskMovements = new TreeSet<>(
+            (movement, other) -> {
+                final int numCaughtUpClients = tasksToCaughtUpClients.get(movement.task).size();
+                final int otherNumCaughtUpClients = tasksToCaughtUpClients.get(other.task).size();
+                if (numCaughtUpClients != otherNumCaughtUpClients) {
+                    return numCaughtUpClients - otherNumCaughtUpClients;
```

Comment:
It's just my dumb little heuristic for hopefully producing a less unbalanced assignment, but I'll add test coverage. Gotta make sure it's not _too_ dumb

Reply:
By the way, my personal take is that we shouldn't worry too much about the intermediate assignment being unbalanced. But, if we do decide this could/should be improved, or get complains from users, I wrote up a quick brainstorm on a hopefully-less-dumb algorithm on the ticket

Reply:
Sounds good. I have no concern about the heuristic.

About the anti-pattern, I was specifically talking about subtracting two integers as a Comparator#compare return value, though. The recommended pattern is to return one of `-1`, `0`, and `1`.

About the lack of test coverage, the concern isn't that we return a less-balanced assignment, but that we break everything completely (somehow). If it's worth writing the code, it's worth verifying it works as expected.

Reply:
Oh I did fix it to use `Integer.compare`, forgot to respond to that part of the comment. Will push the latest changes soon

Reply:
Ah, okeydoke.